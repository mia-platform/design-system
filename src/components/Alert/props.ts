/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ReactNode } from 'react'

import { IconComponent } from '../Icon/Icon.props.ts'
import { Type } from './types.ts'

export type AlertProps = {

  /**
   * The title displayed at the top of the alert.
   */
  title?: ReactNode

  /**
   * The description providing additional details about the alert.
   */
  description?: ReactNode

  /**
   * Custom children content to be rendered inside the alert.
   */
  children?: ReactNode

  /**
   * Indicates if the alert is closable.
   */
  isClosable?: boolean

  /**
   * The type of alert.
   */
  type?: Type

  /**
   * If true, renders a compact version of the alert with reduced padding.
   */
  isCompressed?: boolean

  /**
   * A custom icon to display in the alert. Replaces the default icon based on the `type`.
   */
  icon?: IconComponent

  /**
   * A custom action element (e.g., a button) to display inside the alert.
   */
  action?: ReactNode

  /**
   * A callback function triggered when the alert is closed.
   */
  onClose?: () => void
}
