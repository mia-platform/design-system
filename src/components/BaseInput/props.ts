import { InputRef } from 'antd'
import { Ref } from 'react'

import { Appearance } from './types.ts'

export type BaseInputProps = {

  /**
   * The placeholder for the input.
   */
  placeholder?: string

  /**
   * The type of the Input.
   */
  appearance?: Appearance

  /**
   * Whether the input is disabled.
   */
  isDisabled?: boolean,

  /**
   * Whether the input has error.
   */
  isError?: boolean

  /**
   * Whether the input is read-only.
   */
  isReadOnly?: boolean

  /**
   * Whether the input is full width. Defaults to true.
   */
  isFullWidth?: boolean

  /**
   * Ref to the antd input.
   */
  inputRef?: Ref<InputRef>

  /**
   * Classes for the input.
   */
  className?: string

}
