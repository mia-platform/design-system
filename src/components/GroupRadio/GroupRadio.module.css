.options {
  display: flex;
  flex-direction: column;
  gap: var(--spacing-gap-lg, 16px)
}