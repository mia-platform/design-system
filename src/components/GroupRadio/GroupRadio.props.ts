import { RadioChangeEvent } from 'antd'

export type GroupRadioOption = {
  label: string;
  description?: string;
  value: string | number | boolean;
  disabled?: boolean;
  title?: string;
  id?: string;
};

export type GroupRadioChangeEvent = {
  value: string | number | boolean | undefined;
  event?: RadioChangeEvent;
};

export type GroupRadioProps = {
  defaultValue?: string | number | boolean;
  disabled?: boolean;
  options: GroupRadioOption[];
  onChange?: (changeEvent: GroupRadioChangeEvent) => void;
};
