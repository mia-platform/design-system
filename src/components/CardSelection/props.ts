/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ReactNode } from 'react'

import { InputType, Layout } from './types.ts'
import { IconComponent } from '../Icon/Icon.props.ts'

export type CardSelectionOption<T> = {

  /**
   * Optional id value for the card.
   */
  id?: string

  /**
   * The main title displayed prominently on the card.
   */
  title: ReactNode

  /**
   * An optional icon displayed alongside the title.
   */
  icon?: IconComponent

  /**
   * A subtitle displayed below the title.
   */
  subtitle?: ReactNode

  /**
   * Custom content to be rendered inside the card.
   */
  content?: ReactNode

  /**
   * The value associated with the card.
   */
  value?: T

  /**
   * Determines if the card is disabled.
   */
  disabled?: boolean
}

export type CardSelectionProps<T> = {

  /**
   * The default value for the card selection.
   * Must be an array if inputType is "checkbox", a value if inputType is "radio".
   */
  defaultValue?: T | T[]

  gap?: number

  /**
   * The inputType of the card selection. Defaults: undefined (no input will be displayed).
   */
  inputType?: InputType,

  isDisabled?: boolean

  /**
   * The layout of the cards. Can be "horizontal" or "vertical". Default: "vertical".
   */
  layout: Layout

  /*
  * Handler triggered when the value of the selection changes (only triggered if inputType is "checkbox" or "radio")
  * */
  onChange?: (value: T | T[]) => void

  /*
  * Handler triggered when a card is clicked
  * */
  onClick?: (value?: T) => void

  /**
   * The options array for the card selection.
   */
  options: CardSelectionOption<T>[]

  /**
   * The controlled value for the card selection.
   * Must be an array if inputType is "checkbox", a value if inputType is "radio".
   */
  value?: T | T[]
}
