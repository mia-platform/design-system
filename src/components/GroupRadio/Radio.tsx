import { Radio as AntRadio } from 'antd'
import { ReactElement } from 'react'

import { BodyS } from '../Typography/BodyX/BodyS'
import { RadioProps } from './Radio.props'
import styles from './Radio.module.css'

const { radio, radioContent, radioLabel, radioDescription } = styles

export const Radio = ({
  label,
  description,
  disabled,
  value,
}: RadioProps): ReactElement => {
  return (
    <div className={radio}>
      <AntRadio disabled={disabled} value={value}>
        <div className={radioContent}>
          <div className={radioLabel}>
            <BodyS>{label}</BodyS>
          </div>
          {description && (
            <div className={radioDescription}>
              <BodyS>{description}</BodyS>
            </div>
          )}
        </div>
      </AntRadio>
    </div>
  )
}
