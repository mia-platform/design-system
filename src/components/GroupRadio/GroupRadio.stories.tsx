import { Meta, StoryObj } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import { GroupRadioOption, GroupRadioProps } from './GroupRadio.props'
import { GroupRadio } from './GroupRadio'

const baseOptions: GroupRadioOption[] = [
  {
    label: 'Option 1',
    value: 1,
  },
  {
    label: 'Option 2',
    value: 2,
  },
]

const optionsWithDescription: GroupRadioOption[] = [
  {
    label: 'Option 1',
    value: 1,
    description: 'Description 1',
  },
  {
    label: 'Option 2',
    value: 2,
    description: 'Description 2',
  },
]

const optionsPartiallyDisabled: GroupRadioOption[] = [
  {
    label: 'Option 1',
    value: 1,
    description: 'Description 1',
    disabled: true,
  },
  {
    label: 'Option 2',
    value: 2,
    description: 'Description 2',
  },
]

const allOptionsDisabled: GroupRadioOption[] = [
  {
    label: 'Option 1',
    value: 1,
    description: 'Description 1',
    disabled: true,
  },
  {
    label: 'Option 2',
    value: 2,
    description: 'Description 2',
    disabled: true,
  },
]

const getArgs = (
  options: GroupRadioOption[],
  defaultValue?: number,
  disabled = false,
): GroupRadioProps => ({
  options,
  defaultValue,
  disabled,
  onChange: action('onChange'),
})

const meta = {
  component: GroupRadio,
  args: getArgs(baseOptions),
} satisfies Meta<typeof GroupRadio>

export default meta

type Story = StoryObj<typeof meta>;

export const WithoutDefaultValue: Story = {}

export const WithDefaultValue: Story = {
  args: getArgs(baseOptions, 2),
}

export const WithNonExistentDefaultValue: Story = {
  args: getArgs(baseOptions, 3),
}

export const WithDescription: Story = {
  args: getArgs(optionsWithDescription),
}

export const Disabled: Story = {
  args: getArgs(baseOptions, undefined, true),
}

export const PartiallyDisabled: Story = {
  args: getArgs(optionsPartiallyDisabled),
}

export const PartiallyDisabledWithDefault: Story = {
  args: getArgs(optionsPartiallyDisabled, 1),
}

export const AllOptionsDisabledWithDefault: Story = {
  args: getArgs(allOptionsDisabled, 1),
}

export const AllOptionsDisabledWithoutDefault: Story = {
  args: getArgs(allOptionsDisabled),
}

