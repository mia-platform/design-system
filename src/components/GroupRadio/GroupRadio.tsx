import { Radio as AntRadio, RadioChangeEvent } from 'antd'
import { ReactElement, ReactNode, useCallback, useEffect, useMemo, useState } from 'react'

import { GroupRadioProps } from './GroupRadio.props'
import { Radio } from './components/Radio'
import styles from './GroupRadio.module.css'

const { options: optionsClass, groupRadio } = styles

const defaults = {
  disabled: false,
}

export const GroupRadio = ({
  defaultValue,
  disabled = defaults.disabled,
  options,
  onChange,
}: GroupRadioProps): ReactElement => {
  const [value, setValue] = useState<string | number | boolean | undefined>()

  const computeInitialValue = useCallback(() => {
    let initialValue = options.find(option => option.value === defaultValue)?.value
    if (defaultValue === undefined) {
      for (const option of options) {
        if (!option.disabled) {
          initialValue = option.value
          break
        }
      }
    }
    return initialValue
  }, [defaultValue, options])

  useEffect(() => {
    const initialValue = computeInitialValue()
    setValue(initialValue)
    if (onChange) {
      onChange({ value: initialValue })
    }
  }, [computeInitialValue, onChange])

  const radioOptions = useMemo((): ReactNode => {
    return options.map((option) => (
      <div key={option.value.toString()}>
        <Radio
          checked={option.value === value}
          description={option.description}
          disabled={option.disabled}
          label={option.label}
          value={option.value}
        />
      </div>
    ))
  }, [options, value])

  const handleChange = (event: RadioChangeEvent): void => {
    setValue(event.target.value)
    if (onChange) {
      onChange({ value: event.target.value, event })
    }
  }

  return (
    <div className={groupRadio}>
      <AntRadio.Group
        className={groupRadio}
        disabled={disabled}
        value={value}
        onChange={handleChange}
      >
        <div className={optionsClass}>{radioOptions}</div>
      </AntRadio.Group>
    </div>
  )
}
