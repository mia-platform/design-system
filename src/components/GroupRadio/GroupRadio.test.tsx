import { render, screen, waitFor } from '@testing-library/react'

import { GroupRadioOption, GroupRadioProps } from './GroupRadio.props'
import { GroupRadio } from './GroupRadio'
import { userEvent } from '../../test-utils'

const baseOptions: GroupRadioOption[] = [
  {
    value: 1,
    label: 'option 1',
  },
  {
    value: 2,
    label: 'option 2',
  },
]

const baseProps: GroupRadioProps = {
  defaultValue: 1,
  options: baseOptions,
}

describe('GroupRadio', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('renders correctly with base props', () => {
    const { asFragment } = render(<GroupRadio {...baseProps} />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders correctly with base props', () => {
    const { asFragment } = render(<GroupRadio {...baseProps} />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders correctly with base props and change selected radio', async() => {
    const onChange = jest.fn()
    const props: GroupRadioProps = {
      ...baseProps,
      onChange,
    }

    render(<GroupRadio {...props} />)

    const radios = screen.getAllByRole('radio')
    expect(radios).toHaveLength(2)
    await waitFor(() => expect(onChange).toHaveBeenCalled())
    expect(onChange).toHaveBeenCalledTimes(1)
    const [[invocation]] = onChange.mock.calls
    expect(invocation.event).toBeUndefined()
    expect(invocation.value).toEqual(1)
    onChange.mockClear()

    const secondRadio = screen.getByRole('radio', { name: /option 2/ })
    userEvent.click(secondRadio)
    await waitFor(() => expect(onChange).toHaveBeenCalled())
    expect(onChange).toHaveBeenCalledTimes(1)
    const [[invocation1]] = onChange.mock.calls
    expect(invocation1.event.target.value).toEqual(2)
    expect(invocation1.value).toEqual(2)
  })

  it('cannot select any options if the component is disabled', async() => {
    const onChange = jest.fn()
    const props: GroupRadioProps = {
      ...baseProps,
      disabled: true,
      onChange,
    }

    render(<GroupRadio {...props} />)

    const firstRadio = screen.getByRole('radio', { name: /option 1/ })
    const secondRadio = screen.getByRole('radio', { name: /option 2/ })
    await waitFor(() => expect(onChange).toHaveBeenCalled())
    expect(onChange).toHaveBeenCalledTimes(1)
    const [[invocation]] = onChange.mock.calls
    expect(invocation.event).toBeUndefined()
    expect(invocation.value).toEqual(1)

    onChange.mockClear()

    userEvent.click(firstRadio)
    await waitFor(() => expect(onChange).not.toHaveBeenCalled())
    userEvent.click(secondRadio)
    await waitFor(() => expect(onChange).not.toHaveBeenCalled())
  })

  it('can select only non disabled options', async() => {
    const options: GroupRadioOption[] = [
      {
        value: 1,
        label: 'option disabled',
        disabled: true,
      },
      {
        value: 2,
        label: 'option enabled',
      },
    ]
    const onChange = jest.fn()
    const props: GroupRadioProps = {
      defaultValue: 1,
      options,
      onChange,
    }

    render(<GroupRadio {...props} />)

    const enabeldRadio = screen.getByRole('radio', { name: /option enabled/ })
    const disabledRadio = screen.getByRole('radio', { name: /option disabled/ })
    userEvent.click(enabeldRadio)
    await waitFor(() => expect(onChange).toHaveBeenCalled())
    expect(onChange).toHaveBeenCalledTimes(1)
    onChange.mockClear()
    userEvent.click(disabledRadio)
    await waitFor(() => expect(onChange).not.toHaveBeenCalled())
  })
})
