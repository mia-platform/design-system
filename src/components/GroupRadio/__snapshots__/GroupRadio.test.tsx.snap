// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupRadio renders correctly with base props 1`] = `
<DocumentFragment>
  <div
    class="groupRadio"
  >
    <div
      class="ant-radio-group ant-radio-group-outline groupRadio css-dev-only-do-not-override-5wsri9"
    >
      <div
        class="options"
      >
        <div>
          <div
            class="radio"
          >
            <label
              class="ant-radio-wrapper ant-radio-wrapper-checked css-dev-only-do-not-override-5wsri9"
            >
              <span
                class="ant-radio ant-wave-target ant-radio-checked"
              >
                <input
                  class="ant-radio-input"
                  type="radio"
                  value="1"
                />
                <span
                  class="ant-radio-inner"
                />
              </span>
              <span>
                <div
                  class="radioContent"
                >
                  <div
                    class="radioLabel"
                  >
                    <div
                      class="ant-typography bodyS css-dev-only-do-not-override-5wsri9"
                      role="paragraph"
                    >
                      option 1
                    </div>
                  </div>
                </div>
              </span>
            </label>
          </div>
        </div>
        <div>
          <div
            class="radio"
          >
            <label
              class="ant-radio-wrapper css-dev-only-do-not-override-5wsri9"
            >
              <span
                class="ant-radio ant-wave-target"
              >
                <input
                  class="ant-radio-input"
                  type="radio"
                  value="2"
                />
                <span
                  class="ant-radio-inner"
                />
              </span>
              <span>
                <div
                  class="radioContent"
                >
                  <div
                    class="radioLabel"
                  >
                    <div
                      class="ant-typography bodyS css-dev-only-do-not-override-5wsri9"
                      role="paragraph"
                    >
                      option 2
                    </div>
                  </div>
                </div>
              </span>
            </label>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`GroupRadio renders correctly with base props 2`] = `
<DocumentFragment>
  <div
    class="groupRadio"
  >
    <div
      class="ant-radio-group ant-radio-group-outline groupRadio css-dev-only-do-not-override-5wsri9"
    >
      <div
        class="options"
      >
        <div>
          <div
            class="radio"
          >
            <label
              class="ant-radio-wrapper ant-radio-wrapper-checked css-dev-only-do-not-override-5wsri9"
            >
              <span
                class="ant-radio ant-wave-target ant-radio-checked"
              >
                <input
                  class="ant-radio-input"
                  type="radio"
                  value="1"
                />
                <span
                  class="ant-radio-inner"
                />
              </span>
              <span>
                <div
                  class="radioContent"
                >
                  <div
                    class="radioLabel"
                  >
                    <div
                      class="ant-typography bodyS css-dev-only-do-not-override-5wsri9"
                      role="paragraph"
                    >
                      option 1
                    </div>
                  </div>
                </div>
              </span>
            </label>
          </div>
        </div>
        <div>
          <div
            class="radio"
          >
            <label
              class="ant-radio-wrapper css-dev-only-do-not-override-5wsri9"
            >
              <span
                class="ant-radio ant-wave-target"
              >
                <input
                  class="ant-radio-input"
                  type="radio"
                  value="2"
                />
                <span
                  class="ant-radio-inner"
                />
              </span>
              <span>
                <div
                  class="radioContent"
                >
                  <div
                    class="radioLabel"
                  >
                    <div
                      class="ant-typography bodyS css-dev-only-do-not-override-5wsri9"
                      role="paragraph"
                    >
                      option 2
                    </div>
                  </div>
                </div>
              </span>
            </label>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
